File: AI Bots/botPersonality.ts
// ============================================================================
//  botPersonality.ts  —  Estemshan Intelligent Bot System
//  PHASE 3a:  Bot Personalities + Adaptive Difficulty
// ----------------------------------------------------------------------------
//  This is the "make it #1" layer. Two ideas no normal card app ships well:
//
//    1. PERSONALITIES — each bot has a recognisable style (aggressive,
//       conservative, trickster). Players remember "that bot always Dashes".
//       This is pure flavour ON TOP of the existing skill tiers: a bot can be
//       EXPERT+Aggressive or EASY+Conservative independently.
//
//    2. ADAPTIVE DIFFICULTY — silently keep the player in the fun zone. If they
//       lose repeatedly, ease off; if they steamroll, harden. The single best
//       retention lever in any single-player game.
//
//  Self-contained: imports only ./types and ./botEngine. Additive — nothing
//  breaks if you don't wire it.
// ============================================================================

import { CardModel, Suit } from './types';
import { BotTier, BotBidResult, evaluateBotBid } from './botEngine';
import { estimateBidBySimulation, estimateDashSuccess } from './botSimulation';

// ----------------------------------------------------------------------------
//  1. PERSONALITIES
// ----------------------------------------------------------------------------
export type Personality = 'BALANCED' | 'AGGRESSIVE' | 'CONSERVATIVE' | 'TRICKSTER';

export interface PersonalityProfile {
  key: Personality;
  /** Added to the raw trick estimate before rounding. + = bids higher. */
  bidBias: number;
  /** Scales how readily the bot declares a Dash. >1 = dashes more often. */
  dashEagerness: number;
  /** 0..1 — appetite for pushing a bid to a Super Call (8+) when close. */
  superCallAppetite: number;
  /** Player-facing label + one-line flavour. */
  displayName: string;
  blurb: string;
}

export const PERSONALITIES: Record<Personality, PersonalityProfile> = {
  BALANCED: {
    key: 'BALANCED',
    bidBias: 0,
    dashEagerness: 1.0,
    superCallAppetite: 0.3,
    displayName: 'Steady',
    blurb: 'Plays it straight. Bids what the hand is worth.',
  },
  AGGRESSIVE: {
    key: 'AGGRESSIVE',
    bidBias: +0.4,        // trimmed from +0.8 — a nudge, not a trap
    dashEagerness: 0.6,
    superCallAppetite: 0.7,
    displayName: 'The Shark',
    blurb: 'Bids a touch high, hunts Super Calls, pressures the table.',
  },
  CONSERVATIVE: {
    key: 'CONSERVATIVE',
    bidBias: -0.5,        // slightly softened
    dashEagerness: 1.6,
    superCallAppetite: 0.1,
    displayName: 'The Fox',
    blurb: 'Underbids for safety and loves a clean Dash.',
  },
  TRICKSTER: {
    key: 'TRICKSTER',
    bidBias: +0.15,       // trimmed from +0.3 — variance comes from Dashes, not overbidding
    dashEagerness: 1.4,
    superCallAppetite: 0.6,
    displayName: 'The Joker',
    blurb: 'High variance — unpredictable Dashes and bold calls.',
  },
};

/**
 * Apply a personality to a base bid. Skill (tier) decides HOW WELL the bot
 * evaluates; personality decides its STYLE on top of that evaluation.
 */
export function applyPersonalityToBid(
  base: BotBidResult,
  personality: Personality,
  callerBid: number = -1,
): BotBidResult {
  const p = PERSONALITIES[personality];
  // A conservative bot is more willing to keep a Dash; an aggressive one less so.
  // We re-decide Dash using the personality's eagerness against a soft threshold.
  let isDashCall = base.isDashCall;
  if (base.isDashCall && p.dashEagerness < 1.0) {
    // Aggressive types sometimes convert a marginal Dash into a real bid.
    if (base.rawFloat * (1 / p.dashEagerness) >= 3) isDashCall = false;
  }
  let bid = base.bid;
  if (!isDashCall) {
    let adjusted = base.rawFloat + p.bidBias;
    bid = Math.round(adjusted);
    // Super-call appetite: if close to 8 and appetite is high, push for it.
    if (bid >= 7 && bid < 8 + 0 && p.superCallAppetite >= 0.6 && base.rawFloat >= 7.0) {
      bid = 8;
    }
    // Clamp only to [0..13]. The minimum-4 auction rule is NOT applied here —
    // the caller treats a bid < 4 as a PASS in the auction, and uses it directly
    // as a low estimate in the Estimation phase. (Forcing 4 made weak hands bid
    // unmakeable contracts.)
    if (bid < 0) bid = 0;
    if (bid > 13) bid = 13;
    if (callerBid >= 0 && bid > callerBid) bid = callerBid; // respect Call Cap
  } else {
    bid = 0;
  }
  return {
    ...base,
    bid,
    isDashCall,
    // Recompute — personality can change bid/isDashCall above, and this is a
    // derived field, not something that should carry over stale from `base`.
    isCallPhaseLegal: isDashCall || bid >= 4,
    reasoning: `${base.reasoning} | personality=${p.displayName}(${p.key}) -> bid ${isDashCall ? 'DASH' : bid}`,
  };
}

/**
 * Convenience: evaluate a bid with BOTH skill tier and personality in one call.
 */
export function evaluateBotBidWithPersonality(
  hand: CardModel[],
  tier: BotTier,
  personality: Personality,
  options: {
    otherBids?: number[];
    callerBid?: number;
    playerId?: string;
    chosenTrump?: Suit | 'NONE' | 'POTENTIAL';
    mandatoryTrump?: Suit | 'NONE' | null;
  } = {},
): BotBidResult {
  let base = evaluateBotBid(hand, tier, options);
  // HARD and EXPERT use SIMULATION for both the Dash decision and the bid number.
  if (tier === 'EXPERT' || tier === 'HARD') {
    const samples = tier === 'EXPERT' ? 150 : 100; // HARD a touch lighter
    // 1) VERIFY a heuristic Dash by simulating its real success probability.
    //    Expected-tricks is only a proxy; a hand rated ~0.5 can still be forced
    //    to win a couple of tricks. If P(win 0) is too low, cancel the Dash.
    if (base.isDashCall) {
      const dashP = estimateDashSuccess(hand, { samples });
      const threshold = tier === 'EXPERT' ? 0.85 : 0.80;
      if (dashP < threshold) {
        base = {
          ...base,
          isDashCall: false,
          reasoning: base.reasoning + ` | dash REJECTED (P(0)=${(dashP * 100).toFixed(0)}% < ${threshold * 100}%)`,
        };
      } else {
        base = { ...base, reasoning: base.reasoning + ` | dash OK (P(0)=${(dashP * 100).toFixed(0)}%)` };
      }
    }
    // 2) If not (or no longer) dashing, bid the trick count it can most reliably
    //    make under the assumed trump. (Essential for HARD: at full confidence it
    //    trusted the optimistic heuristic and over-bid. EXPERT still pulls ahead
    //    via its Monte-Carlo CARD PLAY. MEDIUM/EASY keep the heuristic on purpose.)
    if (!base.isDashCall) {
      const assumedTrump: Suit | 'NONE' =
        options.chosenTrump && options.chosenTrump !== 'POTENTIAL'
          ? options.chosenTrump
          : base.potentialTrump;
      const sim = estimateBidBySimulation(hand, assumedTrump, { samples });
      base = {
        ...base,
        rawFloat: sim.bid,
        intendedBid: sim.bid,
        bid: sim.bid,
        isCallPhaseLegal: base.isDashCall || sim.bid >= 4,
        reasoning: base.reasoning + ` | ${tier} sim-bid=${sim.bid} (modal P=${(sim.confidence * 100).toFixed(0)}%)`,
      };
    }
  }
  return applyPersonalityToBid(base, personality, options.callerBid ?? -1);
}

// ----------------------------------------------------------------------------
//  2. ADAPTIVE DIFFICULTY
// ----------------------------------------------------------------------------
//  Track the human's recent results and nudge bot strength to keep games close.
//  Store `recentHumanWins` as a rolling window of booleans (true = human won
//  the round / placed top). Keep this in your game state or localStorage.
// ----------------------------------------------------------------------------

const TIER_ORDER: BotTier[] = ['EASY', 'MEDIUM', 'HARD', 'EXPERT'];

function shift(tier: BotTier, delta: number): BotTier {
  const i = TIER_ORDER.indexOf(tier);
  const j = Math.max(0, Math.min(TIER_ORDER.length - 1, i + delta));
  return TIER_ORDER[j];
}

/**
 * Given the base tier and the human's recent win-rate (0..1 over the last N
 * rounds), return the tier the bot should actually use this round.
 * 
 * winRate < 0.25  -> ease off one tier (player is struggling)
 * winRate > 0.75  -> harden one tier (player is dominating)
 * otherwise        -> keep base tier (good balance)
 * 
 * `floorTier`/`ceilTier` let you clamp so e.g. an EXPERT seat never drops below
 * HARD, or an EASY tutorial bot never rises above MEDIUM.
 */
export function adaptTier(
  baseTier: BotTier,
  humanRecentWinRate: number,
  bounds: { floor?: BotTier; ceil?: BotTier } = {},
): BotTier {
  let t = baseTier;
  if (humanRecentWinRate < 0.25) t = shift(t, -1);
  else if (humanRecentWinRate > 0.75) t = shift(t, +1);
  if (bounds.floor && TIER_ORDER.indexOf(t) < TIER_ORDER.indexOf(bounds.floor)) t = bounds.floor;
  if (bounds.ceil && TIER_ORDER.indexOf(t) > TIER_ORDER.indexOf(bounds.ceil)) t = bounds.ceil;
  return t;
}

/** Maintain a rolling window of the human's recent round outcomes. */
export function pushHumanResult(window: boolean[], humanWonRound: boolean, size = 6): boolean[] {
  const next = [...window, humanWonRound];
  return next.slice(-size);
}

export function winRate(window: boolean[]): number {
  if (window.length === 0) return 0.5; // neutral until we have data
  return window.filter(Boolean).length / window.length;
}

// ----------------------------------------------------------------------------
//  3. SUGGESTED TABLE SETUP
// ----------------------------------------------------------------------------
//  A ready-made trio of opponents that feels varied and memorable. Assign these
//  at game init alongside each bot's skill tier.
// ----------------------------------------------------------------------------
export interface BotIdentity {
  tier: BotTier;
  personality: Personality;
}

export const SUGGESTED_TABLE: Record<string, BotIdentity> = {
  p2: { tier: 'EASY',   personality: 'CONSERVATIVE' }, // "The Fox" — gentle, cautious
  p3: { tier: 'HARD',   personality: 'AGGRESSIVE'   }, // "The Shark" — real pressure
  p4: { tier: 'EXPERT', personality: 'TRICKSTER'    }, // "The Joker" — wildcard genius
};

// ============================================================================
//  WIRING
// ----------------------------------------------------------------------------
//  • Add optional fields to PlayerModel:
//       personality?: Personality;
//       baseTier?: DifficultyLevel;     // if you want adaptive to remember the seat's base
//  • At bid time, instead of evaluateBotBid(...), call:
//       evaluateBotBidWithPersonality(hand, tier, bot.personality ?? 'BALANCED', {...})
//  • For adaptive difficulty, keep a rolling window in game state:
//       humanResults = pushHumanResult(humanResults, humanPlacedTopThisRound);
//    then each round compute the live tier:
//       const liveTier = adaptTier(bot.baseTier, winRate(humanResults),
//                                  { floor: 'EASY', ceil: 'EXPERT' });
//  • Show PERSONALITIES[p].displayName + blurb in the UI next to each bot for
//    instant character (e.g. "The Shark 🦈").
// ============================================================================

File: native/engine/src/main/java/com/estemshan/engine/bot/BotPersonality.kt
package com.estemshan.engine.bot

/**
 * Bot personalities (Module E — the "style" layer), ported from
 * `AI Bots/botPersonality.ts` §1.
 *
 * **Personality is style, [BotTier] is skill, and the two are orthogonal.**
 * The tier decides *how well* a bot evaluates a hand (see [HandEvaluator] and
 * [BidBrain]); the personality decides what it *does* with that evaluation —
 * an EXPERT+Aggressive and an EASY+Aggressive both nudge their estimate the
 * same direction, they simply arrive at the number with different accuracy.
 * Any seat can be any pairing, exactly like the source's
 * "a bot can be EXPERT+Aggressive or EASY+Conservative independently".
 *
 * The TypeScript models this as a `Record<Personality, PersonalityProfile>`
 * lookup table keyed by a string union. The port folds the table into enum
 * constructor parameters, exactly like [BotTier] did for `TIER_CONFIG`: same
 * values, idiomatic Kotlin, and the profile can never drift out of sync with
 * the enum's cases the way a parallel table can. The source's redundant `key`
 * field is the enum constant's own name and is therefore not ported.
 *
 * **These numbers are load-bearing.** They were tuned against real bot-vs-bot
 * sessions in the JS app — the source's own comments record the values they
 * were trimmed from (`AGGRESSIVE.bidBias` was +0.8, `TRICKSTER`'s +0.3) and
 * why: a nudge, not a trap, because "variance comes from Dashes, not
 * overbidding". The golden personality tests (S9) pin the *direction* each
 * profile moves a hand, and S10's bot-vs-bot smoke test pins a whole match, so
 * a retune here shows up as a failing test rather than a silent behaviour
 * change. They are not up for adjustment in this port.
 *
 * **Not ported here, by design.** The source's `evaluateBotBidWithPersonality`
 * is folded into [BidBrain.decide] so a personality-aware bid still has to clear
 * `canSubmit` like any other intent; `adaptTier` / `pushHumanResult` /
 * `winRate` (adaptive difficulty) are v1.1; `SUGGESTED_TABLE` belongs to the
 * Choose Level lobby, a later story.
 */
enum class BotPersonality(
  /** Added to the raw trick estimate *before* rounding. Positive = bids
   * higher. This shifts the pre-round float, not the rounded bid, so a bias
   * smaller than the distance to the next integer is a genuine no-op on a
   * given hand — which is why the golden tests assert a *direction* across a
   * hand chosen to sit on a rounding boundary, not an exact delta everywhere. */
  val bidBias: Double,
  /** How readily the bot declares a Dash. **>1 = dashes more often.** The
   * source applies this as a threshold multiplier on the heuristic dash signal
   * rather than a probability: a conservative profile (1.6) keeps a marginal
   * dash that an aggressive one (0.6) converts back into a real bid. Applied
   * in [BidBrain]'s dash decision. */
  val dashEagerness: Double,
  /** 0..1 — appetite for pushing a bid to a Super Call (8+) when the estimate
   * already sits at 7. The source only fires above 0.6, so CONSERVATIVE (0.1)
   * effectively never super-calls and AGGRESSIVE (0.7) almost always does. */
  val superCallAppetite: Double,
  /** Player-facing label for the lobby and the seat roster. */
  val displayName: String,
  /** One-line flavour text for the UI; no game effect. */
  val blurb: String,
) {
  BALANCED    ( 0.00, 1.0, 0.3, "Steady",    "Plays it straight. Bids what the hand is worth."),
  AGGRESSIVE  ( 0.40, 0.6, 0.7, "The Shark", "Bids a touch high, hunts Super Calls, pressures the table."),
  CONSERVATIVE(-0.50, 1.6, 0.1, "The Fox",   "Underbids for safety and loves a clean Dash."),
  TRICKSTER   ( 0.15, 1.4, 0.6, "The Joker",  "High variance — unpredictable Dashes and bold calls."),
  ;

  companion object {
    /** The source's implicit default when a seat has no personality assigned
     * (`bot.personality ?? 'BALANCED'`); used as the default value wherever a
     * personality is injected. */
    val DEFAULT: BotPersonality = BALANCED
  }
}

/**
 * Apply a personality's style to a tier-skill bid estimate — the port of
 * `botPersonality.ts`'s `applyPersonalityToBid`, adapted to [BidBrain]'s
 * [BotBid]. Pure: same estimate + same personality always yields the same bid,
 * with no state, no randomness, and no knowledge of the auction around it.
 *
 * Skill decides how well the hand was evaluated; this decides what the bot
 * *does* with that number. Two effects, both straight from the source:
 *
 *  * **[BotPersonality.bidBias]** shifts the estimate before the single
 *    rounding point, so an aggressive bot bids a touch high and a conservative
 *    one a touch low.
 *  * **[BotPersonality.superCallAppetite]** pushes a hand already sitting at 7
 *    up to the 8-trick Super Call, but only for personalities above the
 *    source's 0.6 appetite line and only when the raw estimate itself reached
 *    7 — a bias that merely rounded a 6.8 up to 7 does not qualify.
 *
 * **Two deliberate divergences from the source**, both documented in the
 * file they touch:
 *
 *  * The source applies its bias to the *un-jittered* float, which silently
 *    discards the tier's seat jitter and would make every same-personality
 *    seat bid identically. This port applies the bias to the jittered
 *    [BotBid.noisy], so style lands on top of jitter rather than replacing it.
 *  * The source's `callerBid` clamp (never bid above the Caller's cap) is not
 *    ported. In the port, `canSubmit` and the per-phase folds are the single
 *    authority on auction legality — the engine already rejects a bid above
 *    the cap, so re-clamping here would be a second, drift-prone copy of the
 *    rule.
 *
 * The [BotPersonality.DEFAULT] personality is the identity: it returns [base]
 * untouched, which is what keeps every pre-S9 caller of `BidBrain` behaving
 * byte-identically.
 */
internal fun applyPersonalityToBid(base: BotBid, personality: BotPersonality): BotBid {
  if (personality == BotPersonality.DEFAULT) return base

  val biased = base.noisy + personality.bidBias
  var intendedBid = kotlin.math.round(biased).toInt()

  var superCall = false
  if (intendedBid == SUPER_CALL_TRIGGER &&
    personality.superCallAppetite >= SUPER_CALL_APPETITE_GATE &&
    base.rawFloat >= SUPER_CALL_RAW_FLOOR
  ) {
    intendedBid = SUPER_CALL
    superCall = true
  }

  val clamped = intendedBid.coerceIn(0, 13)
  return base.copy(
    intendedBid = clamped,
    reasoning = base.reasoning +
      " | personality=${personality.displayName}(${personality.name}) -> bid $clamped" +
      (if (superCall) " (super call)" else ""),
  )
}

private const val SUPER_CALL_TRIGGER = 7
private const val SUPER_CALL = 8
private const val SUPER_CALL_APPETITE_GATE = 0.6
private const val SUPER_CALL_RAW_FLOOR = 7.0
